urllib3/util/hpkp.py

```diff
@@ -0,0 +1,269 @@
+        self.hosts = {}
+
+    def store_host(self, known_host):
+        self.hosts[known_host.domain] = known_host
+
+    def invalidate_host(self, domain):
+        del self.hosts[domain]
+
+    def iter_hosts(self):
+        return self.hosts.values()
+
+
+class HPKPManager(object):
+    """
+    An object that manages HPKP for urllib3.
+
+    This object is responsible for validating HPKP for HTTPS connections. It
+    employs a :class:`HPKPDatabase <urllib3.util.hpkp.HPKPDatabase>` object to
+    manage the actual storage, but the logic is contained in this class.
+    """
+    def __init__(self, database):
+        self.db = database
+
+    def validate_connection(self, domain, socket):
+        """
+        For a given connection, confirm that it passes HPKP validation.
+
+        This checks whether there are any ``KnownPinnedHost``s for the given
+        connection. If they are, validates that the presented certificate
+        actually matches the pin.
+        """
+        hosts = self.db.get_hosts_by_domain(domain)
+        if not hosts:
+            # This connection has not previously done any key pinning, so
+            # there's no validation to do.
+            return
+
+        host = self._find_valid_pinned_host(domain, hosts)
+        if host is None:
+            # Again, no previous pinning was done on this connection, so
+            # there's no validation to do.
+            return
+
+        # If we got this far, we have a KnownPinnedHost in hand that applies to
+        # this connection. Get the certificate and check it matches one of the
+        # pins.
+        # TODO: Actually do this!
+        return
+
+    def process_response(self, domain, response):
+        """
+        Processes a HTTPS response for HPKP.
+
+        This method checks whether the PKP header is present on the request,
+        and if it is updates its store of key pins. It also validates the
+        connection against the HPKP header, confirming that it does pass at
+        this stage.
+        """
+        # TODO: Handle ip address domains.
+        # If no PKP header is provided, then no action is required.
+        pkp = response.getheader('public-key-pins')
+        if not pkp:
+            return
+
+        # Try to get the pins out. If this fails for any reason, abort the
+        # pinning process.
+        try:
+            pin = parse_public_key_pins(pkp, domain)
+        except HPKPError:
+            return
+
+        # Quick check: if the max-age is 0, or the expiry date is after now,
+        # we should invalidate any pin we already have for this domain and
+        # exit.
+        if not pin.max_age or (pin.max_age + pin.start_date <= time.time()):
+            self.db.invalidate_host(domain)
+            return
+
+        # Otherwise, we can store off this pin.
+        self.db.store_host(pin)
+
+        # TODO: Check if the RFC wants us to immediately validate the
+        # connection: if it does, do it.
```

Comment:
Yeah, this is the bit I actually needed from the spec:

>  - The TLS connection was authenticated with a certificate chain containing at least one of the SPKI structures indicated by at least one of the given SPKI Fingerprints (see Section 2.6).
> - The given set of Pins contains at least one Pin that does NOT refer to an SPKI in the certificate chain.  (That is, the host must set a Backup Pin; see Section 4.3.)

We need to do both of these checks here.